src/libgrate/grate-3d-ctx.c

```diff
@@ -423,70 +423,90 @@ void grate_3d_ctx_set_provoking_vtx_last(struct grate_3d_ctx *ctx, bool last)
 	ctx->provoking_vtx_last = last;
 }
 
-struct grate_texture * grate_3d_ctx_create_texture(struct grate_3d_ctx *ctx,
-						   unsigned width,
-						   unsigned height,
-						   unsigned wrap_mode,
-						   unsigned max_lod,
-						   unsigned format,
-						   bool mip_filter,
-						   bool mag_filter,
-						   bool min_filter,
-						   struct host1x_bo *tex_bo)
-{
-	struct grate_texture *tex;
-
-	switch (format) {
-	case PIXEL_FORMAT_A8:
-	case PIXEL_FORMAT_L8:
-	case PIXEL_FORMAT_S8:
-	case PIXEL_FORMAT_LA88:
-	case PIXEL_FORMAT_RGB565:
-	case PIXEL_FORMAT_RGBA5551:
-	case PIXEL_FORMAT_RGBA4444:
-	case PIXEL_FORMAT_D16_LINEAR:
-	case PIXEL_FORMAT_D16_NONLINEAR:
-	case PIXEL_FORMAT_RGBA8888:
-	case PIXEL_FORMAT_RGBA_FP32:
-		break;
-	default:
-		grate_error("Invalid format %u\n", format);
-		return NULL;
+int grate_3d_ctx_activate_texture(struct grate_3d_ctx *ctx, unsigned location)
+{
+	if (location >= 16) {
+		grate_error("Invalid location %u\n", location);
+		return -1;
 	}
 
-	tex = calloc(1, sizeof(*tex));
-	if (!tex) {
-		grate_error("Allocation failed\n");
-		return NULL;
+	ctx->active_texture = location;
+
+	return 0;
+}
+
+void grate_3d_ctx_bind_texture(struct grate_3d_ctx *ctx,
+			       struct host1x_pixelbuffer *pb)
+{
+	ctx->textures[ctx->active_texture].pb = pb;
+}
+
+int grate_3d_ctx_set_texture_max_lod(struct grate_3d_ctx *ctx,
+				     unsigned location,
+				     unsigned max_lod)
+{
+	if (location >= 16) {
+		grate_error("Invalid location %u\n", location);
+		return -1;
 	}
 
-	tex->bo = tex_bo;
-	tex->format = format;
-	tex->max_lod = max_lod;
-	tex->wrap_mode = wrap_mode;
-	tex->width = width;
-	tex->height = height;
-	tex->mip_filter = mip_filter;
-	tex->mag_filter = mag_filter;
-	tex->min_filter = min_filter;
+	ctx->textures[ctx->active_texture].max_lod = max_lod;
```

Comment:
While this is the way OpenGL used to do things like this, bindless is now the new wine, where the texture object is specified directly. And especially for a low-level API, I think the following makes a lot more sense; perhaps we make it up to the callee to keep track of what object they want to modify (e.g pass `struct grate_texture *` to these functions instead of `struct grate_3d_ctx *`)? That would surely lead to less boiler-plate code...